File: haskell/010/euler.hs
import Data.List

main :: IO ()

sieve :: (Integral a) => a -> a -> [a] -> [a] -> [a]
sieve prime limit primes integers
    | prime > limit = sort ((prime:primes) ++ integers)
    | otherwise     = sieve newPrime limit (prime:primes) newIntegers
                      where
                        (newPrime:newIntegers) = filter (\x -> mod x prime /= 0) integers

sieveOfEratosthenes :: (Integral a) => a -> [a]
sieveOfEratosthenes limit = sieve 2 (round $ sqrt $ fromIntegral $ limit) [] [3, 5..limit]

main = print(sum (sieveOfEratosthenes 2000000))

File: erlang/010/euler.erl
#!/usr/bin/env escript

main(_) ->
  io:fwrite("~w~n", [lists:foldl(fun(X, Sum) -> X + Sum end, 0, sieve_of_eratosthenes(2000000))]).

sieve_of_eratosthenes(Limit) when is_number(Limit), Limit > 1 ->
  sieve_of_eratosthenes(2, round(math:sqrt(Limit)), [], lists:seq(3, Limit, 2)).
sieve_of_eratosthenes(Prime, Limit, Primes, Integers) when Prime > Limit ->
  lists:reverse([Prime | Primes]) ++ Integers;
sieve_of_eratosthenes(Prime, Limit, Primes, Integers) ->
  [NewPrime | NewIntegers] = lists:filter(fun(X) -> X rem Prime /= 0 end, Integers),
  sieve_of_eratosthenes(NewPrime, Limit, [Prime | Primes], NewIntegers).
